build: clang fails because of -Wnull-pointer-subtraction

darthunix opened the pull request:
The problem was detected on MacOS llvm with `-DCMAKE_BUILD_TYPE=Debug` option. Tarantool fails to build because of `-Wnull-pointer-subtraction` warning in `EIGHT_BYTE_ALIGNMENT` macros.
```
/Users/darthunix/git/tarantool/src/box/sql/malloc.c:125:9: error: performing pointer subtraction with a null pointer has undefined behavior [-Werror,-Wnull-pointer-subtraction]
        assert(EIGHT_BYTE_ALIGNMENT(p));        /* IMP: R-11148-40995 */
               ^~~~~~~~~~~~~~~~~~~~~~~
/Users/darthunix/git/tarantool/src/box/sql/sqlInt.h:839:49: note: expanded from macro 'EIGHT_BYTE_ALIGNMENT'
#define EIGHT_BYTE_ALIGNMENT(X)   ((((char*)(X) - (char*)0)&7)==0)
                                                ^ ~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/assert.h:99:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __ASSERT_FILE_NAME, __LINE__, #e) : (void)0)
                        ^
/Users/darthunix/git/tarantool/src/box/sql/malloc.c:232:9: error: performing pointer subtraction with a null pointer has undefined behavior [-Werror,-Wnull-pointer-subtraction]
        assert(EIGHT_BYTE_ALIGNMENT(pNew));     /* IMP: R-11148-40995 */
               ^~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/darthunix/git/tarantool/src/box/sql/sqlInt.h:839:49: note: expanded from macro 'EIGHT_BYTE_ALIGNMENT'
#define EIGHT_BYTE_ALIGNMENT(X)   ((((char*)(X) - (char*)0)&7)==0)
                                                ^ ~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/assert.h:99:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __ASSERT_FILE_NAME, __LINE__, #e) : (void)0)
                        ^
2 errors generated.
```
Fixed.

CuriousGeorgiy commented on src/box/sql/sqlInt.h:
```diff
@@ -834,9 +834,9 @@ typedef u64 uptr;
  */
 #ifdef SQL_4_BYTE_ALIGNED_MALLOC
-#define EIGHT_BYTE_ALIGNMENT(X)   ((((char*)(X) - (char*)0)&3)==0)
+#define EIGHT_BYTE_ALIGNMENT(X)   ((((uintptr_t)((char*)(X)) - (uintptr_t)((char*)0))&3)==0)
```
```suggestion
#define EIGHT_BYTE_ALIGNMENT(X)   ((((uintptr_t)((char*)(X)) & 3) == 0)
```

CuriousGeorgiy commented on src/box/sql/sqlInt.h:
```diff
@@ -834,9 +834,9 @@ typedef u64 uptr;
+#define EIGHT_BYTE_ALIGNMENT(X)   ((((uintptr_t)((char*)(X)) - (uintptr_t)((char*)0))&3)==0)
 #else
-#define EIGHT_BYTE_ALIGNMENT(X)   ((((char*)(X) - (char*)0)&7)==0)
+#define EIGHT_BYTE_ALIGNMENT(X)   ((((uintptr_t)((char*)(X)) - (uintptr_t)((char*)0))&7)==0)
```
```suggestion
#define EIGHT_BYTE_ALIGNMENT(X)   ((((uintptr_t)((char*)(X)) & 7) == 0)
```

darthunix replied:
Honestly, at first I also wondered about the reason to subtract the NULL address in the original macros. But it seems that some machines can have non-zero address of the NULL pointer.

darthunix replied:
So, if you think that we never execute Tarantool on such strange machines, I'll remove subtraction of a pointer with a zero address.

CuriousGeorgiy replied:
> But it seems that some machines can have non-zero address of the NULL pointer.

Could you provide a reference please?

CuriousGeorgiy replied:
> So, if you think that we never execute Tarantool on such strange machines, I'll remove subtraction of a pointer with a zero address.
According to our [README](https://github.com/tarantool/tarantool/blob/master/README.md):
https://github.com/tarantool/tarantool/blob/7aa02f302d819727f68e8f455ea23a066a00b4b5/README.md?plain=1#L41-L42

CuriousGeorgiy replied:
AFAIC, this subtraction was needed for an implicit conversion to [`ptrdiff_t`](https://en.cppreference.com/w/c/types/ptrdiff_t) without including the `stddef.h` header file.

darthunix replied:
> Furthermore, the macros subtracted literal 0, not NULL

If I understand the non-zero NULL pointer architectures correctly (I have never seen them in reality, only googled), this statement is incorrect. The macros subtracted an address on the null pointer as `(char*)0` that returns `x0` on regular machines and some non-zero value on non-zeroed machines. It is compiler's responsibility to understand that `0` is a NULL pointer and translate it correctly to non-zero bytes.

CuriousGeorgiy replied:
> So, I believe this macro has been inherited from the SQLite that decided to double check such an exotic case. It compares the distance in bytes between our address and the address of the NULL pointer (that can be non-zero value).

I really doubt this was the semantics that the authors implied. Judging by the name of the macro, I believe the purpose of this subtraction is the following:
> AFAIC, this subtraction was needed for an implicit conversion to [ptrdiff_t](https://en.cppreference.com/w/c/types/ptrdiff_t) without including the stddef.h header file.

CuriousGeorgiy replied:
As far as I can trace back the source code history, the purpose of this code was really only to check alignment — not any kind of 'distance comparison' that you mentioned:
https://github.com/sqlite/sqlite/commit/c96d8530ca54c7ec7c6ca52b9e7f01f33f0c4235

https://github.com/sqlite/sqlite/commit/66e8008404d70107795683f0fa2b533e0e641f46

darthunix commented:
I can see that the linter test failed. What linter do you use in `checkpatch/checkpatch.pl --color=always --git HEAD~1..HEAD` script?

darthunix commented:
@CuriousGeorgiy I have simplified the macro as we have discussed. Also changed comment message in a way `checkpatch`  expects it (but pleas, take a look whether it is ok).

CuriousGeorgiy commented on changelogs/unreleased/build-clang-fails-because-of-null-pointer-subtraction.md:
```diff
@@ -0,0 +1,3 @@
+## bugfix/build
```
We do have a `build` subsystem, but AFAI, it only includes CMake and CI/CD scripts.

The subsystem you patch here is simply called `sql`. Same comment regarding your commit message subject.

Nit: your commit message is quite formal — please provide some context, e.g.:

```git
sql: fix -Wnull-pointer-subtraction warning

clang 13 includes a check for subtraction from `NULL` pointer which is considered UB: historically, we had an alignment checking macro which is affected by this. It seems like the intention of the macro's author was to implicitly cast the pointer being checked to `uintptr_t` without including `stddef.h` — replace this subtraction with an explicit cast.
```

CuriousGeorgiy commented on changelogs/unreleased/build-clang-fails-because-of-null-pointer-subtraction.md:
```diff
@@ -0,0 +1,3 @@
+## bugfix/build
+
+* Fix clang errors on debug builds caused by an incorrect NULL pointer subtraction in assertion macros.
```
Actually, I don't think this changelog is necessary: this bug could not be observed by users — please drop it.

CuriousGeorgiy reviewed (approved):
Please address my comment about your commit message and drop the changelog — otherwise, LGTM.

ImeevMA commented on src/box/sql/sqlInt.h:
```diff
@@ -834,9 +834,9 @@ typedef u64 uptr;
  */
 #ifdef SQL_4_BYTE_ALIGNED_MALLOC
-#define EIGHT_BYTE_ALIGNMENT(X)   ((((char*)(X) - (char*)0)&3)==0)
+#define EIGHT_BYTE_ALIGNMENT(X)   ((((uintptr_t)((char *)(X))) & 3) == 0)
```
We have way to set `SQL_4_BYTE_ALIGNED_MALLOC`
```Diff
diff --git a/src/box/sql/sqlInt.h b/src/box/sql/sqlInt.h
index 8e83743a6..07497f18d 100644
--- a/src/box/sql/sqlInt.h
+++ b/src/box/sql/sqlInt.h
@@ -828,16 +828,8 @@ typedef u64 uptr;
  * Assert that the pointer X is aligned to an 8-byte boundary.  This
  * macro is used only within assert() to verify that the code gets
  * all alignment restrictions correct.
- *
- * Except, if sql_4_BYTE_ALIGNED_MALLOC is defined, then the
- * underlying malloc() implementation might return us 4-byte aligned
- * pointers.  In that case, only verify 4-byte alignment.
  */
-#ifdef SQL_4_BYTE_ALIGNED_MALLOC
-#define EIGHT_BYTE_ALIGNMENT(X)   ((((uintptr_t)((char *)(X))) & 3) == 0)
-#else
 #define EIGHT_BYTE_ALIGNMENT(X)   ((((uintptr_t)((char *)(X))) & 7) == 0)
-#endif
 
 /*
  * Default maximum size of memory used by memory-mapped I/O in the VFS
```

darthunix replied:
Done

ImeevMA reviewed (changes requested):
Hi! Thank you for the patch. I have only one comment.

ImeevMA reviewed (approved):
Thank you for the fixes! LGTM

darthunix commented:
I can see that 25 tests failed (same amount as I can see in https://github.com/tarantool/tarantool/pull/7067 ). It seems that CI is partially broken. Can you rerun it after CI been fixed?

ImeevMA commented:
I'm not sure what will help, most likely both PRs need to be rebased to the current master. Could you rebase both PRs on the current master?

Totktonada commented:
Cherry-picked to 2.10 in 2.10.0-2-g969cd6800.
